File: Assets/Typescript~/src/Shared/UserInput/Generic/Pointer.ts
import { Bin } from "Shared/Util/Bin";
import { Touchscreen } from "../Touchscreen";
import { Mouse } from "../Mouse";
import { Signal } from "Shared/Util/Signal";

export class Pointer {
	private readonly bin = new Bin();
	private readonly touchscreen = new Touchscreen();
	private readonly mouse = new Mouse();

	public readonly Down = new Signal<void>();
	public readonly Up = new Signal<void>();
	public readonly Moved = new Signal<[location: Vector3]>();

	constructor() {
		this.bin.Add(this.touchscreen);
		this.bin.Add(this.mouse);
		this.bin.Add(this.mouse.LeftDown.Proxy(this.Down));
		this.bin.Add(this.mouse.LeftUp.Proxy(this.Up));
		this.bin.Add(this.mouse.Moved.Proxy(this.Moved));
		this.bin.Connect(this.touchscreen.PrimaryTouch, (location, phase) => {
			switch (phase) {
				case TouchPhase.Began:
					this.Down.Fire();
					break;
				case TouchPhase.Ended:
					this.Up.Fire();
					break;
				case TouchPhase.Moved:
					this.Moved.Fire(location);
					break;
			}
		});
		this.bin.Add(this.Down);
		this.bin.Add(this.Up);
		this.bin.Add(this.Moved);
	}

	/**
	 * Cleans up the pointer listeners.
	 */
	public Destroy() {
		this.bin.Destroy();
	}
}

File: Assets/Bundles/Shared/Resources/TS/UserInput/Generic/Pointer.lua
-- Compiled with unity-ts v2.1.0-75
local Bin = require("Shared/TS/Util/Bin").Bin
local Touchscreen = require("Shared/TS/UserInput/Touchscreen").Touchscreen
local Mouse = require("Shared/TS/UserInput/Mouse").Mouse
local Signal = require("Shared/TS/Util/Signal").Signal
local Pointer
do
	Pointer = setmetatable({}, {
		__tostring = function()
			return "Pointer"
		end,
	})
	Pointer.__index = Pointer
	function Pointer.new(...)
		local self = setmetatable({}, Pointer)
		return self:constructor(...) or self
	end
	function Pointer:constructor()
		self.bin = Bin.new()
		self.touchscreen = Touchscreen.new()
		self.mouse = Mouse.new()
		self.Down = Signal.new()
		self.Up = Signal.new()
		self.Moved = Signal.new()
		self.bin:Add(self.touchscreen)
		self.bin:Add(self.mouse)
		self.bin:Add(self.mouse.LeftDown:Proxy(self.Down))
		self.bin:Add(self.mouse.LeftUp:Proxy(self.Up))
		self.bin:Add(self.mouse.Moved:Proxy(self.Moved))
		self.bin:Connect(self.touchscreen.PrimaryTouch, function(location, phase)
			repeat
				if phase == 1 then
					self.Down:Fire()
					break
				end
				if phase == 3 then
					self.Up:Fire()
					break
				end
				if phase == 2 then
					self.Moved:Fire(location)
					break
				end
			until true
		end)
		self.bin:Add(self.Down)
		self.bin:Add(self.Up)
		self.bin:Add(self.Moved)
	end
	function Pointer:Destroy()
		self.bin:Destroy()
	end
end
return {
	Pointer = Pointer,
}
-- ----------------------------------
-- ----------------------------------
